NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,532 @@
+		try
+		{
+			result += fs::file_size(entry.path());
+		}
+		catch (fs::filesystem_error& e)
+		{
+			if (entry.is_directory())
+			{
+				result += GetSizeOfFolderContentsMinusFile(entry.path(), "");
+			}
+		}
+	}
+	return result;
+}
+
+// Saves a file asynchronously.
+template <ScriptContext context> void SaveFileManager::SaveFileAsync(fs::path file, std::string contents)
+{
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread writeThread(
+		[mutex, file, contents]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				fs::path dir = file.parent_path();
+				// this actually allows mods to go over the limit, but not by much
+				// the limit is to prevent mods from taking gigabytes of space,
+				// we don't need to be particularly strict.
+				if (GetSizeOfFolderContentsMinusFile(dir, file.filename().string()) + contents.length() > MAX_FOLDER_SIZE)
+				{
+					// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
+					spdlog::error(fmt::format("Mod spamming save requests? Folder limit bypassed despite previous checks. Not saving."));
+					mutex.get().unlock();
+					return;
+				}
+
+				std::ofstream fileStr(file);
+				if (fileStr.fail())
+				{
+					mutex.get().unlock();
+					return;
+				}
+
+				fileStr.write(contents.c_str(), contents.length());
+				fileStr.close();
+
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("SAVE FAILED!");
+				mutex.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	writeThread.detach();
+}
+
+// Loads a file asynchronously.
+template <ScriptContext context> int SaveFileManager::LoadFileAsync(fs::path file)
+{
+	int handle = ++m_iLastRequestHandle;
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread readThread(
+		[mutex, file, handle]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				std::ifstream fileStr(file);
+				if (fileStr.fail())
+				{
+					spdlog::error("A file was supposed to be loaded but we can't access it?!");
+
+					g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, false, "");
+					mutex.get().unlock();
+					return;
+				}
+
+				std::stringstream stringStream;
+				while (fileStr.peek() != EOF)
+					stringStream << (char)fileStr.get();
+
+				g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, true, stringStream.str());
+
+				fileStr.close();
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
```

Comment:
If we make concurrent accesses fail instead of waiting, we can reduce the size of the critical section and make a global map of strings to bools protected by a single mutex instead. This will prevent buggy mods from queuing infinitely, while also resolving the leak.

Reply:
That seems unnecessarily complex - we could just use a simple mutex for all file actions to avoid the leak completely (since an SSD/hard drive can only read one file at once anyways)

I don't think concurrent reads from multiple mods should result in a fail, as that would just make file reads on startup fail 99% of the time, and would complicate the file reading process too much.

Reply:
> we could just use a simple mutex for all file actions to avoid the leak completely (since an SSD/hard drive can only read one file at once anyways)

That may be suitable, but we'd have to test it. I'd be happy with that solution for now, though, and to deal with it if it causes performance issues later.

> concurrent reads from multiple mods should result in a fail

I mean concurrent accesses to the same file, and specifically, concurrent writes/deletes while reading. Multiple reads are fine.

Reply:
> I'd be happy with that solution for now, though, and to deal with it if it causes performance issues later.

So this is resolved?

Reply:
bump

Reply:
this is resolved. we now use a single mutex for all file operations.